src/transformStateWithClones.js

```diff
@@ -6,8 +6,32 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  let resultObject = { ...state };
+  const resultArray = [];
+
+  for (let i = 0; i < actions.length; i++) {
+    if (actions[i].type === 'addProperties') {
+      resultArray.push(Object.assign({}, resultObject, actions[i].extraData));
+      Object.assign(resultObject, actions[i].extraData);
+    }
 
+    if (actions[i].type === 'removeProperties') {
+      for (const key of actions[i].keysToRemove) {
+        if (resultObject.hasOwnProperty(key)) {
+          delete resultObject[key];
+        }
+      }
+      resultArray.push({ ...resultObject });
+    }
+
+    if (actions[i].type === 'clear') {
+      resultArray.push({});
+      resultObject = {};
+    }
```

Comment:
Good work but let's rewrite it using switch/case because we rely on a single value `actions[i].type`